feat: add type coverage

saimeunt opened the pull request:
## Motivation and Resolution

This PR adds a new `ts:coverage` npm script that uses [typescript-coverage-report](https://www.npmjs.com/package/typescript-coverage-report) to generage a TypeScript coverage report available in the coverage-ts directory.

On initial run, every files had >80% coverage except 4:
`src/utils/assert.ts`
`src/utils/calldata/formatter.ts`
`src/utils/calldata/parser/parser-2.0.0.ts`
`src/utils.responseParser/index.ts`

I added some additional typings on these files to reach >80% coverage (not sure if these modifications are 100% correct though).

Resolves: #1088 

### RPC version (if applicable)

0.7.1

## Usage related changes

<!-- How the changes from this PR affect users. -->

None

## Development related changes

New npm script added: `ts:coverage`

## Checklist:

- [X] Performed a self-review of the code
- [X] Rebased to the last commit of the target branch (or merged it into my branch)
- [X] Linked the issues which this PR resolves
- [ ] Documented the changes in code (API docs will be generated automatically)
- [ ] Updated the tests
- [X] All tests are passing

ivpavici reviewed (commented):
thanks! 
we were thinking about removing this whole `coverage-ts` folder , and instead just provide the option for developers to locally check if the type coverage is ok in the console... and this could be also added to our CI on every PR. So thinking about
https://github.com/plantain-00/type-coverage instead of https://github.com/alexcanessa/typescript-coverage-report

saimeunt commented:
@ivpavici I removed the `coverage-ts` folder and switched to `type-coverage` as requested.
The `ts:coverage` command will now display a warning eg. `The type coverage rate is lower than the target(95%).`
The 95% target is arbitrary idk if we should target higher or lower.
Also I couldn't find a way with this package to display precisely the specific files having a low type coverage contrary to `typescript-coverage-report`.

ivpavici commented:
thanks!
Maybe we couldn also add a note for future devs about running this step manually here:
https://github.com/starknet-io/starknet.js/blob/develop/CONTRIBUTING.md#other-notes

PhilippeR26 commented on src/utils/responseParser/index.ts:
```diff
@@ -11,13 +13,13 @@ import {
 
 export abstract class ResponseParser {
-  abstract parseGetBlockResponse(res: any): GetBlockResponse;
+  abstract parseGetBlockResponse(res: BlockWithTxHashes): GetBlockResponse;
```
I have a serious doubt for this one.
@tabaktoni  should have a look at all the types you added.

PhilippeR26 reviewed (changes requested):
The global ts coverage percentage is nice to have, but if the result is bad, the dev has nothing to identify what and where are the problems.
The report is anyway useful if the percentage has to be improved. 
Maybe a npm command just to get the result value, but also a command to have the report if requested by the dev. Will have to be explained (also how to display the report) in CONTRIBUTE doc.

saimeunt commented:
@PhilippeR26 Thanks for the review, I added [typescript-coverage-report](https://www.npmjs.com/package/typescript-coverage-report) back and ignored the default output `coverage-ts` in `.gitignore`.
@ivpavici I've added a small note about the 2 new npm scripts introduced in CONTRIBUTING.md

PhilippeR26 reviewed (changes requested):
- After analysis, I think that the types added are good, but has to be confirmed by @tabaktoni .

- Script for coverage report is generating a full screen warning :
```
Warning: Table: Support for defaultProps will be removed from function components in a future major release. Use JavaScript default parameters instead.
    at Table (/home/edmond/Documents/starknet/starknet.js/node_modules/semantic-ui-react/dist/commonjs/collections/Table/Table.js:40:24)
    at div
    at Container 
...
```
Can you solve that?

- You should add in Contribute (otherwise sure it will be asked in the assistance ...) something like : `launch ./coverage-ts/index.html in your browser.`

- Target your PR to `next-version`

tabaktoni commented:
In general looking lgtm

ivpavici commented:
@saimeunt can you please fix the last comments from Phil? and we can merge 👍

saimeunt commented:
Will do tonight!

saimeunt commented:
@PhilippeR26 I was unable to make the warning disappear, it's caused by a non-critical issue in semantic-ui-react which is a dependency of typescript-coverage-report, this dependency will have to be updated to get rid of the issue.

@ivpavici I'm closing this PR in favor of #1120 which is properly targetting next-version.
